### ISSUE TITLE
DataProducerPluginBase::resolve() should exist and be abstract

### ISSUE THREAD

**Issue by chx** [drupal-graphql/graphql#861]
Right now resolve is enforced to exist via a method_exist call, it should be an abstract method instead. And I get the variable arguments problem but perhaps an array or a ParameterBag if we want fancy would be a good solution?

**Comment by fubhy** [comment]
I agree. This has been discussed during the last refactoring of the plugins already but I didn't want to add another BC breaking change in the same alpha release. Since it's a convenience method that one can circumvent entirely by directly implementing `resolveField` instead and reading the contexts from the plugin instance manually I am a still unsure what the best approach would be. I copied this pattern from the rules module where they have a similar problem with their action / condition plugins and solve it exactly the same way.

**Comment by klausi** [comment]
I think the method_exists() call is fine because we want to pass the contexts as direct parameters to the resolve() method. A ParameterBag or something would be quite a heavy API change at this point and make the code on a data producer less clear. Getting in the context with type hints is really useful, I would like to keep that. Even if that means not being able to express the resolve() method on the interface.
